packages/std/src/coins.rs

```diff
@@ -0,0 +1,336 @@
+use std::collections::BTreeMap;
+use std::fmt;
+use std::str::FromStr;
+
+use crate::{errors::CoinsError, Coin, StdError, StdResult, Uint128};
+
+/// A collection of coins, similar to Cosmos SDK's `sdk.Coins` struct.
+///
+/// Differently from `sdk.Coins`, which is a vector of `sdk.Coin`, here we
+/// implement Coins as a BTreeMap that maps from coin denoms to amounts.
+/// This has a number of advantages:
+///
+/// - coins are naturally sorted alphabetically by denom
+/// - duplicate denoms are automatically removed
+/// - cheaper for searching/inserting/deleting: O(log(n)) compared to O(n)
+#[derive(Clone, Default, Debug, PartialEq, Eq)]
+pub struct Coins(BTreeMap<String, Uint128>);
+
+// Casting a Vec<Coin> to Coins.
+// The Vec can be out of order, but must not contain duplicate denoms or zero amounts.
+impl TryFrom<Vec<Coin>> for Coins {
+    type Error = CoinsError;
+
+    fn try_from(vec: Vec<Coin>) -> Result<Self, CoinsError> {
+        if let Some(coin) = vec.iter().find(|c| c.amount.is_zero()) {
+            return Err(CoinsError::ZeroAmount {
+                denom: coin.denom.clone(),
+            });
+        }
+
+        let vec_len = vec.len();
+
+        let map = vec
+            .into_iter()
+            .filter(|coin| !coin.amount.is_zero())
+            .map(|coin| (coin.denom, coin.amount))
+            .collect::<BTreeMap<_, _>>();
+
+        // the map having a different length from the vec means the vec must either contain
+        // duplicate denoms
+        if map.len() != vec_len {
+            return Err(CoinsError::DuplicateDenom);
+        }
+
+        Ok(Self(map))
+    }
+}
+
+impl TryFrom<&[Coin]> for Coins {
+    type Error = CoinsError;
+
+    fn try_from(slice: &[Coin]) -> Result<Self, CoinsError> {
+        slice.to_vec().try_into()
+    }
+}
+
+impl<const N: usize> TryFrom<[Coin; N]> for Coins {
+    type Error = CoinsError;
+
+    fn try_from(slice: [Coin; N]) -> Result<Self, CoinsError> {
+        slice.to_vec().try_into()
+    }
+}
+
+impl From<Coins> for Vec<Coin> {
+    fn from(value: Coins) -> Self {
+        value.into_vec()
+    }
+}
+
+impl FromStr for Coins {
+    type Err = StdError;
+
+    fn from_str(s: &str) -> StdResult<Self> {
+        if s.is_empty() {
+            return Ok(Self::default());
+        }
```

Comment:
👍 Interesting behaviour by split here:

```
let v: Vec<&str> = "".split('X').collect();
assert_eq!(v, [""]);
```

I thought an empty collection was the default.